libs/core/algorithms/tests/unit/algorithms/util/test_is_trivially_relocatable.cpp

```diff
@@ -0,0 +1,55 @@
+//  Copyright (c) 2023 Isidoros Tsaousis-Seiras
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/algorithms/traits/is_trivially_relocatable.hpp>
+
+#include <cassert>
+
+// Trivially Copyable types are trivially relocatable
+struct TriviallyCopyable
+{
+    TriviallyCopyable(TriviallyCopyable const&) = default;
+};
+static_assert(hpx::is_trivially_relocatable_v<TriviallyCopyable>,
+    "Trivially Copyable type should be Trivially Relocatable");
+
+// Non trivially copyable types are not trivially relocatable
+// Unless they are explicitly declared as such
+struct NotTriviallyCopyable_1
+{
+    NotTriviallyCopyable_1(NotTriviallyCopyable_1 const&){};
+};
+static_assert(!hpx::is_trivially_relocatable_v<NotTriviallyCopyable_1>,
+    "Not Trivially Copyable and Not declared Trivially Relocatable type should "
+    "not be Trivially Relocatable");
```

Comment:
I see, what could a trivially relocatable type do in the assignment operator to break swap?

Reply:
Literally anything other than "assign the value of the right-hand side to the left-hand side." I'm actually planning a blog post for this week or next, on all the things that potentially go wrong when we treat `std::pmr::string` as trivially relocatable.

```
struct Widget {
    std::pmr::string ps;
};

std::vector<Widget> w;
w.push_back(Widget{ std::pmr::string("hello"), &mr1) });
w.push_back(Widget{ std::pmr::string("world"), &mr2) });
w.erase(w.begin()); // ideally should destroy w[0] and relocate downward, right?
assert(w[0].ps.get_allocator().resource() == &mr1); // but today this is required to be true

auto v1 = std::pmr::vector<std::pmr::string>({"hello", "hello"}, &mr1);
auto v2 = std::pmr::vector<std::pmr::string>({"world", "world"}, &mr2);
std::swap_ranges(v1.begin(), v1.end(), v2.begin()); // ideally should swap the bytes, right?
assert(v1[0].get_allocator().resource() == &mr1); // but today this is required to be true
```

There are several theories of what to do here:
- Don't call `std::pmr::string` trivially relocatable at all. (This is the default, implied by the current wording of P1144R8.)
- Call `std::pmr::string` trivially relocatable, and put a precondition on _all_ standard library facilities that says you're not allowed to pass them objects from different arenas; if you do, that's library UB. (This is my own preferred theory.)
- Call `std::pmr::string` trivially relocatable, and put a precondition on _some_ standard library facilities that says you're not allowed to pass them objects from different arenas; if you do, that's library UB. For example, let's say `vector<Widget>::erase` gets to assume that all the `Widget`s in the vector have the same arena, but `std::swap<Widget&>` does not, and thus simply cannot be optimized for trivially relocatable types. (This is the Bloomberg/P2786R1 theory, as of last week, although it might change in the future.)